hy/models/__init__.py

```diff
@@ -35,3 +35,35 @@ def replace(self, other):
             raise TypeError("Can't replace a non Hy object with a Hy object")
 
         return self
+
+_wrappers = {}
+
+
+def wrap_value(x):
+    """Wrap `x` into the corresponding Hy type.
+
+    This allows replace_hy_obj to convert a non Hy object to a Hy object.
+
+    This also allows a macro to return an unquoted expression transparently.
+
+    """
+
+    wrapper = _wrappers.get(type(x))
+    if wrapper is None:
+        return x
+    else:
+        return wrapper(x)
+
+
+def replace_hy_obj(obj, other):
+
+    if isinstance(obj, HyObject):
+        return obj.replace(other)
+
+    wrapped_obj = wrap_value(obj)
+
+    if isinstance(wrapped_obj, HyObject):
+        return wrapped_obj.replace(other)
+    else:
+        raise TypeError("Can't replace a Hy object "
```

Comment:
This error message doesn't clearly state what's wrong. Maybe we should change it to something like

> "Can't wrap a <type xxx> object to a HyObject"